### ISSUE TITLE
words.protocol - irc.py assumes nickchange to be successful before server ack

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3377]
|<img alt="nazgjunk's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| nazgjunk reported|
|-|-|
|Trac ID|trac#3377|
|Type|defect|
|Created|2008-07-30 22:45:19Z|
|Branch|https://github.com/twisted/twisted/tree/irc-nick-registration-failure-3377|

Before anything, I'd like to apologise for my behaviour in #twisted while stumbling upon this mess.

http://twistedmatrix.com/trac/browser/trunk/twisted/words/protocols/irc.py

IRCClient.setNick() assumes that there will be no problems in setting the new nick. Even before the NICK command is send to the server self.nickname is set to the new nickname, without waiting for the proper server acknowledgement or error message.

In the process I found out that ERR_NICKNAMEINUSE was handled incorrectly - it repeated self.register() which isn't correct, whether while signing on or while just changing nicks.

The enclosed patch should solve everything related to changing/setting nicks and ERR_NICKNAMEINUSE, unless the ircd decides to send error messages at random.

Reasoning for each modification:
Added self._registered: Boolean, depends on whether we're signed in. Needed this to conveniently figure out how to handle ERR_NICKNAMEINUSE. Set to True in self.signedOn()

Edited self.setNick: Don't set self.nickname, but do set self.attemptednick.

Edited self.irc_ERR_NICKNAMEINUSE(): If we're in the process of registering (ircd won't send RPL_WELCOME until we've chosen a nick that is allowed) add an underscore to the end of self.attemptednick and try to self.setNick with that. Don't need to register all over again, the USER line has long been sent and the order of operations while sending in is only a recommendation (RFC2812, Ch. 3.1). If we're already registered and setNick was called by some other mechanism, the user can override the function if he wants to.

This is my first patch to any kind of project not mine ever, and I've only been coding Python for about two months. If there is anything I need to be corrected on, please don't hesitate to pass me a note.

Attachments:

* [irc.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/099/099852c89abee96fcee23c275504ef67203c5317/8f5aa514a1890a453329861e2847c05bedd2a3fa.diff) (1229 bytes) - added by nazgjunk on 2008-07-30 22:46:49Z - diff -u output of patched irc.py
* [latest.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/099/099852c89abee96fcee23c275504ef67203c5317/41e4cef1090fe766355db27613a53c3425e27018.diff) (5024 bytes) - added by nazgjunk on 2008-07-31 22:19:20Z - New diff file from latest revision in trunk - irc.py and test_irc.py
* [irc-nick-changes-latest2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/099/099852c89abee96fcee23c275504ef67203c5317/c9b468127ba1b094a94ee0819d7519b7089e1dd0.patch) (4221 bytes) - added by jesstess on 2009-09-04 18:21:08Z - simplify the nick change process, add a test case for erroneous nicks
* [irc-nick-changes-latest3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/099/099852c89abee96fcee23c275504ef67203c5317/6e5850b454756ae0b22e53bf33e6aab4455238cf.patch) (7079 bytes) - added by jesstess on 2009-09-07 22:02:10Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3377 3377
type__defect defect
reporter__nazgjunk nazgjunk
priority__high high
milestone__ 
branch__branches_irc_nick_registration_failure_3377 branches/irc-nick-registration-failure-3377
branch_author__nazgjunk__jesstess nazgjunk, jesstess
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1217457919000000 1217457919000000
changetime__1252459964000000 1252459964000000
version__None None
owner__ 
cc__exarkun cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix set owner to nazgjunk|
|-|-|

Thank you very much for the patch, nazgjunk.

Before it can be accepted, it'll need unit tests (this is a new policy since most of irc.py was written). I encourage you to try to add some tests to test_irc.py. You'll find some examples in the other tests there.

If you get around to doing that, please add the "review" keyword to this ticket and someone will thoroughly review the patch.

**Comment by twisted-trac** [comment]
|<img alt="nazgjunk's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|nazgjunk commented|
|-|-|

Updated to work with the latest revision of irc.py in trunk, and added a set of NICK related test cases - all OK. Found some bumps I'd forgotten while writing the test cases, fixed those as well. This should handle all NICK interaction with the server fine, except for ERR_NICKCOLLISION since I utterly fail to understand how IRCDs supposedly implement that - the RFC is all but clear.

I added a function to handle ERR_ERRONEUSNICKNAME while registering. While one might say that the user shouldn't try to use a nick full of Unicode, if it does occur the connection will stall if no new NICK is sent which would be horribly confusing to the user.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Hi nazgjunk, thanks for the patch (especially the tests :).

Some hunks in the patch looks like they're not directly related to the changes being made _and_ some of them are changing the code from being coding-standard compliant to not being, so they should be removed:

  * twisted/words/test/test_irc.py
    * changing `ClientTests` from subclassing `TestCase` to subclassing `unittest.TestCase` should be removed - neither form is mandated, but changing from one to the other is just noise in this patch.
    * changing `test_register` to `testRegister` and `test_registerWithPassword` to `testRegisterWithPassword` should be removed - the former is correct in each case according to the coding standard

Regarding the rest of the patch:

  * twisted/words/test/test_irc.py
    * `testRegisterWithTakenNick` should be renamed `test_registerWithTakenNick`.  Likewise, `testNickChange` should be renamed `test_nickChange`.
    * The docstrings for both new test methods are great, although `testRegisterWithTakenNick` talks about something happening while _inside the C{register} method_.  This is a bit misleading, since the `register` method returns before anything actually happens.
  * twisted/words/protocols/irc.py
    * the class-scope initialization of `_attemptedNick` is missing whitespace around `=`
    * ditto for the string concatenation in `irc_ERR_NICKNAMEINUSE`
    * the two new attributes should be documented in the class docstring using `@ivar`
    * The `register` method may have some concurrency issues now.  Calling it twice before a response arrives to the first `NICK` command will have bad consequences.  I'm not sure what the behavior for this case would have been before, but it seems like it probably wouldn't have been as bad.
    * `irc_ERR_NICKNAMEINUSE` will no longer attempt to fix a bad nick set by `setNick` outside of `register`.  This probably isn't a great change.  Why not call `setNick` for the non-registration case as well?
    * Similar comment for `irc_ERR_ERRONEUSNICKNAME`, although skipping the non-registration case  here _may_ not be as bad, since this is a new method so no one could be relying on any old behavior of it.

**Comment by twisted-trac** [comment]
|<img alt="nazgjunk's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|nazgjunk commented|
|-|-|

As for the noisy useless things, erm, yeah. Not quite sure why I did that. Probably wasn't quite awake and trying to get the file more consistent. Will be undone.

 * twisted/words/test/test_irc.py 
   * Same as above.
   * Will have a look at that. I remember that I was trying to avoid saying "while signing on", since that'd be more a reference to the process/protocol than to the function.
 * twisted/words/protocols/irc.py 
   * The documentation in the class docstring - I was under the impression that "private" variables (I'm well aware they're not) weren't documented, but okay.
   * I have no clue as to what you mean. The `register` method is called only once, in `IRCClient.connectionMade`. It shouldn't really be called at any other time. The original behaviour was that on `ERR_NICKNAMEINUSE` it'd be called again, and I removed that. Note that if the NICK command during signon is succesful, the server shouldn't send a reply. 
   * `setNick` doesn't change IRCClient.nickname, intendedly. If it is called outside `register`, the server will send a NICK command to inform us that our nickname has changed, and only then can we start using the new nickname. Changing to a backup nick when `setNick` is called during normal operation doesn't make a lot of sense to me - it just sticks with the old nick, as any normal IRC client I've seen so far would. The only sensible option I see is telling a user that the nickchange was not successful.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess removed owner|
|-|-|

It seems like most of the work has been done on this so I'd like to try and wrap it up.

I went through the latest patch by nazgjunk trying to address exarkun's comments. In doing so, it looked like getting rid of the the _registered checks in exarkun's last two bullets under the twisted/words/protocols/irc.py comments made the attribute unnecessary. Additionally, it looked like self._attemptedNick has the same purpose as self.nickname: the nickname you wish you had whether or not you've been approved by the IRC server, so I think it can be removed as well.

The new patch, latest2, doesn't use _registered or _attemptedNick, makes the cosmetic changes suggested by exarkun, adds a unit test for irc_ERRONEOUSNICKNAME, and adds some markup to the doc strings. It doesn't break any of the old tests and passes the 3 new tests.

I'm not very familiar with irc.py so I may be wrong about the necessity of _attemptedNick. If there's a more comprehensive test case I could write to test the issue let me know.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks for following up on this jesstess.  I think there are a couple things to address in the latest version of the patch:

  1. The `_registered` check may have been important for the `irc_ERR_ERRONEOUSNICKNAME` case.  For both of these error handlers, there are two cases to consider.  Setting the nick during initial sign on (during which time a failure to set the nick is essentially catastrophic to the connection) and setting the nick at any later time (when the failure is probably not extremely harmful).  As implemented now, I think `irc_ERR_ERRONEOUSNICKNAME` mostly makes sense for the former case, but not so much for the latter case.  The nick change attempt should probably just fail for the latter case, and I guess that requires the state `_registered` was tracking in the earlier patches.
  1. The string `'defaultnick'` should probably be a overrideable  somehow.  A class attribute would probably be sensible (ie `IRCClient.ERRONEOUS_NICK_FALLBACK` or some such thing).
  1. I almost agreed with you about `_attemptedNick` not being necessary, but then I realized that there might be multiple failures.  `alice` and `alice_` may both be in use, so the code needs to track what it tried last somehow, otherwise `setNick('alice')` will just try `setNick('alice_')` over and over again.

Ultimately, it would be nice if the API presented for nick management was improved - something like `IRCClient.changeNickname(newNickname)` which returned a `Deferred` that fired with success or failure depending on what came back from the server.  That's likely beyond the scope of this ticket, but I wanted to mention it so that everyone's clear that the current `IRCClient` code is not the ideal, it's just what we're stuck with for now.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess removed owner|
|-|-|

_registered and _attemptedNick are back and documented and joined by erroneousNickDefault. I added a couple of asserts to test_registerWithTakenNick for the multi-collision case. ERR_ERRONEOUSNICKNAME is now tested in both the registered (fall back to old nick) and unregistered (fall back to erroneousNickDefault) cases in test_erroneousNick.

I stuck with the hardcoded '_' for collisions. Opinions on making it overrideable too? I don't think I've ever seen it be anything else.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj set owner to @jesstess|
|-|-|

Great patch.

I just have some minor commentary:

  1. Please can you change `assertEqual` to `assertEquals`.
  2. I see you corrected the spelling of `ERR_ERRONEUSNICKNAME`, which is a bit of a tricky thing. The incorrect spelling actually comes from [the RFC](http://www.mirc.com/help/rfc1459.txt) and has been like this for a while, so there might be existing code that uses that spelling. My suggestion would be to leave the original spelling and create a synonym for the correct spelling; or alternatively just use the wrong spelling, whichever you prefer.
  3. There are a bunch of asserts like `self.assertEqual(lastLine[:4], 'NICK')`, all of which seem to be followed by `self.assertEqual(lastLine, 'NICK therestoftheline')`. Is there any merit in checking the beginning of the line twice?
  4. I think readability can be improved by putting `self.transport.value().split('\r\n')[-2]` into a method (`getLastLine`?)

As for the `_` thing: I've seen some IRC clients do different things (append ```````, replace letters with numeral lookalikes, etc.) Overriding `irc_ERR_NICKNAMEINUSE` to implement your own mechanism isn't rocket science, but on the other hand adding a method that figures out an alternate nickname isn't either. I think this would be a nice addition to this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

+1 on on a _new_ ticket for adding an overrideable method to determine a new nickname to try based on one rejected with an in-use error.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess removed owner|
|-|-|

Points 1-4 by jonathanj are addressed in the new latest3 patch. ERRONEOUS is just back to ERRONEUS.

I opened [#4010](https://github.com/twisted/twisted/issues/4010) for the method to handle nickname collisions.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [27304]) Branching to 'irc-nick-registration-failure-3377'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [27305]) Apply jesstess's modified version of nazgjunk's patch to handle ERR_ERRONEUSNICKNAME (sic) and improve the `nickname` attribute

refs [#3377](https://github.com/twisted/twisted/issues/3377)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [27306]) Add a note about the misspelling of a word, and some whitespace

refs [#3377](https://github.com/twisted/twisted/issues/3377)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Looks good to me.  Thanks for working on this.  I'll merge it now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [27307]) Merge irc-nick-registration-failure-3377

Author: nazgjunk, jesstess
Reviewer: exarkun, jonathanj
Fixes: [#3377](https://github.com/twisted/twisted/issues/3377)

Change `IRCClient` so that `setNick` does not update the `nickname` attribute.
Instead, wait for the server to respond that the nick has actually changed before
updating this attribute.  Also add a handler for the erroneous nickname error
response and have it fall back to a safe name if this happens during registration.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
